Include added mass

joaobrittoneto opened the pull request:
@doudou,
This is how I could include the added mass and complete 6dof parameters to gazebo.
I had to use some workarounds.
By now it depends on base::eigen (matrix6d, vector6d, JacobiSVD). The parameters are pass as string (maybe there is a better way).
Also I include this compensatedEffort, so the engine computes the expected acceleration.
How could I make unit test to validate it?

joaobrittoneto committed:
Include base-types in CMakeList and Eigen in header

Using base::Matrix6d, not defined by gazebo

joaobrittoneto committed:
Workaround for getting matrices parameters

joaobrittoneto committed:
Apply damping as matrices

joaobrittoneto committed:
Apply coriolis of added mass matrix

joaobrittoneto committed:
Apply Coriolis effect of added inertia and compensated Effort

doudou commented on src/GazeboUnderwater.cpp:
```diff
@@ -197,4 +229,137 @@ namespace gazebo_underwater
+
+    base::Matrix6d GazeboUnderwater::mountGzInertiaMatrix(ModelPtr model) const
+    {
+        base::Matrix6d gzInertia = base::Matrix6d::Zero();
```
This should really be using the Gazebo representations ... one can get the inertia information per-link and add them up.

doudou commented:
Would be IMO much better to use gazebo's linear algebra and stick to Gazebo types (e.g. Inertia) (potentially gazebo 7). I would rather keep this package base-types-free if possible to keep its gazebo roots. Moreover, it looks like it would make things simpler (less back-and-forth between gazebo and Eigen)

doudou commented on src/GazeboUnderwater.cpp:
```diff
@@ -115,50 +119,78 @@ namespace gazebo_underwater
+        addedInertia = convertToMatrix(getParameter<string>("added_inertia","Kg, Kg.m2", extra_inertia));
+
+        gzInertia = mountGzInertiaMatrix(model);
+        inverseInertia = gzInertia + addedInertia;
```
Avoid assigning something to a variable that is _not_ what it's meant to hold _in fine_. Just create a temporary completeInertia and inverse it in inverseInertia

doudou commented:
Other question: it seems that gazebo does allow you to access a link's applied force/torques from the previous iteration. Do you think using that would make things better ?

joaobrittoneto commented:
> Would be IMO much better to use gazebo's linear algebra and stick to Gazebo types 

Agreed. As gazebo math doesn't have matrix6d neither vector6d, I used them from base/Eigen.
I can use 4 math::Matrix3 to represent a base::Matrix6d. It may not look mathematically elegant, but can make the code base-types free. (btw, I don't get why gazebo implement it's own math code instead of using a more complete one like eigen)

> Other question: it seems that gazebo does allow you to access a link's applied force/torques from the previous iteration. Do you think using that would make things better ?

I'm not sure. In order to make the added mass be considered by gazebo, I applied this [compensated force](https://github.com/Brazilian-Institute-of-Robotics/simulation-gazebo_underwater/pull/8/commits/84655433967225ac75faa998feca5cbe50f4314b#diff-0ebc57c91816b463bad9d3ef0000761cR264), a trick that makes gazebo computes the right acceleration. 
It takes all forces that will be applied in the present iteration (thruster, gravity term, damping, plus the coriolis of added mass and of the original mass) and add a force "C", in such a way that:
acceleration = M^-1 \* (F+C) = (M+Ma)^-1 \* F 
(M+Ma)^-1 \* F is how gazebo should compute the acceleration considering the added mass.
M^-1 \* F is how it actually does. The "C = (M*(M+Ma)^-1 - I) \* F" is the term to make them equal. 

ps: I'm not sure how to get all forces that will be applied in the present iteration. I did like [that](https://github.com/Brazilian-Institute-of-Robotics/simulation-gazebo_underwater/pull/8/commits/84655433967225ac75faa998feca5cbe50f4314b#diff-0ebc57c91816b463bad9d3ef0000761cR284), is it right?

Other way is to consider the acceleration computed in the past iteration, like this [guy](https://github.com/Mishalassif/gazebo_hydrodynamics/blob/master/src/VirtualMassPlugin.cpp#L196) did:
a[k] = M^-1 \* (F[k] - Ma \* a[k-1] + coriolis(Ma,v))
but there will be an error caused by this step delay.

doudou commented:
> As gazebo math doesn't have matrix6d neither vector6d, I used them from base/Eigen.

It does not, but it does have an explicit representation of the inertial properties (gazebo::physics::Inertial). Might be what you're looking for.

Generally speaking, isn't the 6x6 matrix diagonal by blocks anyways (i.e. only have the upper-left and lower-right 3x3 nonzero) ?

> tw, I don't get why gazebo implement it's own math code instead of using a more complete one like eigen

It's an unfortunate general tendency in the gazebo development team (they basically are spawning their own robotic framework now, yey !).

> ps: I'm not sure how to get all forces that will be applied in the present iteration. I did like that, is it right?

What _is_ sure is that from the point of view of the "age" of the data, it is equivalent to GetWorldLinearAccel() (accel is simply force divided by mass and torque force divided by the inertial matrix). Same for GetWorldAngularAccel(), so using them would not change the properties of the code.

After a cursory look at the gazebo code, I guess that it's indeed one-step-behind data. I would not consider it a problem, honestly. Doing the added mass, even with this one-step time difference, would already improve the simulation's behaviour tremendously IMO.

joaobrittoneto commented:
> Generally speaking, isn't the 6x6 matrix diagonal by blocks anyways (i.e. only have the upper-left and lower-right 3x3 nonzero) ?

I don't want to restrict it. I'd like to make it more generic and consider the case where the 6x6 inertia matrix is full completed, like proposed [here](https://jscholarship.library.jhu.edu/bitstream/handle/1774.2/36985/MCFARLAND-DISSERTATION-2014.pdf) in pag 109 (btw, it's where the 6 6x6 damping matrices come from). For the decoupled case, we just need to set the off diagonal elements to zero.

> What is sure is that from the point of view of the "age" of the data, it is equivalent to GetWorldLinearAccel() 

I was intending to be precise with this compensated force, but due the difficult to get the forces that will be applied, I will use the forces of the added mass with past acceleration. 

Just one question: Does gazebo compute the coriolis effect of the inertia? (I do think it does, just to make sure)

doudou commented:
> [snip]

Fair enough ... The TL;DR is that it's physically not very meaningful, but useful for the AUV "added mass" model, correct ?

> Just one question: Does gazebo compute the coriolis effect of the inertia? (I do think it does, just to make sure)

No idea. You'll have to either dig into the code or ask them I am afraid

joaobrittoneto commented:
Yeah, it represents the coupling terms. It's a little bit hard to visualize and set it by feeling, but I hope future identification of the 6dof coupled model could make the simulation more realistic. Until there we just need to set those coupling terms as zero and we are in the known decoupled case.

Ok, I'll dig it first.
@doudou: Which physics engine do we use in gazebo? Dart, ODE, Bullet, Simbody??

joaobrittoneto commented:
@doudou,
When the migration to gazebo7 will happen? 
The version I have in rock ( gazebo 5.0.1) has some pretty [wrong physics](https://bitbucket.org/osrf/gazebo/src/430e66533f8810646b90d001fc9c83220ae503e9/gazebo/physics/Link.cc?at=gazebo5_5.0.1&fileviewer=file-view-default#Link.cc-616) that are [correct](https://bitbucket.org/osrf/gazebo/src/a1defa74ed98b8dcaff4744e1d5dbdafb8eabbcd/gazebo/physics/Link.cc?at=gazebo7_7.3.0&fileviewer=file-view-default#Link.cc-672) in the new version

quams commented:
I guess it will happen when somebody looks into it and does it... 

What is the effort to migrate to gazebo 7? @doudou any ideas, or a rough evaluation? 

As far as I remember gazebo 4 was the current version when we started the simulation development and Thomio already switched to 5 during that. The list of changes looks pretty extensive. And there seems to be a new Hydrodynamics plugin / underwater vehicle plugin.

doudou commented:
> What is the effort to migrate to gazebo 7? @doudou any ideas, or a rough evaluation?

There's zero effort, I've been running on G7 for months.

What I said on slack a week or two ago is that I would prefer that we migrate our machines to 16.04 first, as G7 is already available as an Ubuntu package there.

quams commented:
Ah, I remember now. I could have sworn I made a trello card for this, with the reason on it. Can't find it anymore.

We should migrate as long as FlatFish is still in parts and we are without contract from Shell.

doudou commented:
Yep. That was my thinking as well. Migrate ASAP, waiting is just more pain

joaobrittoneto committed:
Replace compensated effort to force create by AddedInertia.

Replace base::Eigen by own Matrix6 and Vector6,
making code Gazebo dependent only.

joaobrittoneto commented:
@doudou,

Now the code is Gazebo dependent only. I had to include a Matrix6 and Vector6 based on gazebo's math.

I changed the applyCompensatedEffort by the applyEffortAddedInertia. 
But funny things happen. As the applyEffortAddedInertia use the acceleration of previous step to apply a force, it makes the system becomes unstable, with the acceleration going to infinity.

Short, this solution is not reliable.
IMO, we need to go back to plan A, including the applyCompesatedEffort.
It looks to me that a change on gazebos code need to be made to have access to the forces that will be applied.
I'd need to have access to the [dBodyGetForce](https://bitbucket.org/osrf/gazebo/src/955f398e7d8e949bcb5d7b2094ab9f27d24caeab/deps/opende/src/ode.cpp?at=default&fileviewer=file-view-default#ode.cpp-680), after the thruster efforts are added. The `facc`  and `tacc` are the accumulated forces and torques. Gazebo just [provide](https://bitbucket.org/osrf/gazebo/src/955f398e7d8e949bcb5d7b2094ab9f27d24caeab/gazebo/physics/ode/ODELink.cc?at=default&fileviewer=file-view-default#ODELink.cc-679) the forces [once they are applied](https://bitbucket.org/osrf/gazebo/src/955f398e7d8e949bcb5d7b2094ab9f27d24caeab/gazebo/physics/ode/ODELink.cc?at=default&fileviewer=file-view-default#ODELink.cc-169) on the body. 

What do you think?

joaobrittoneto committed:
Update manifest and CMakeList

doudou commented:
> What do you think?

I think that you might be using the total forces (i.e. force PLUS added mass) in the next step, which would explain why it diverges. No ?

joaobrittoneto commented:
No, no. I paid attention to [that](https://github.com/Brazilian-Institute-of-Robotics/simulation-gazebo_underwater/pull/8/commits/5c4bee515e5ccd0bcc4bd35f982d1e4f3149f7c8#diff-0ebc57c91816b463bad9d3ef0000761cR260).
The instability can be seen with a Z transform on the system:

> m \* a[k] + ma \* a[k-1] = F 
> m \* A(z) + ma \* z⁻¹ \* A(z) = F(z)
> A(z) = F(z)/m / (1 + ma/m \* z⁻¹)

It says that the pole is place at `-ma/m`, being outside the unit circle for the case where `ma > m` and so the system is unstable. 
(It's an analyze for 1dof decoupled system. For the 6dof coupled case, it's more complex).

Running the code we can see that also. For small values of added mass it runs fine. But for big values it crashes, once accelerations go to infinity.  
Since it's normal the added inertia being bigger than the vehicles inertia, I think that solution is not reliable.

joaobrittoneto commented:
How can I make a unit test for the library?
At the moment I need to run the gazebo simulation, syskit and the dive control in 3 terminals in order to have a generic overview. 
I'd like to make it simple and do some math validation.
Any easy way to do that?

doudou commented:
> I'd like to make it simple and do some math validation.

:+1:

> Any easy way to do that?

Not sure. You could have a look at gazebo's own unit tests, I guess. Another way would be to split the math from the pluing itself, so that you can at least test the math.

joaobrittoneto committed:
Define Matrix6 inversion

joaobrittoneto committed:
Update applyDamp

Fix getModelFrameVelocities, remove applyEffortAddedInertia

joaobrittoneto committed:
Add computeModelInertial and update applyBuoyancy

Compute the whole Inertial of the model, including the total mass,
the total moment of inertia and the position of the center of gravity
related to the main link's origin.

joaobrittoneto committed:
compute inverseInertia of model's inertia and addeInertia

joaobrittoneto committed:
Include Matrix6 operator *=, update way of configure matrix

joaobrittoneto committed:
apply added mass compensated efforts

joaobrittoneto committed:
Correct compensated effort

Remove influence of previous compensated effort from previous applied force

joaobrittoneto committed:
Update way of computing compensated efforts

Consider the rigid body Coriolis effect and previous compensated effort.
Update documentation.

joaobrittoneto committed:
Update way of getting matrix parameters

joaobrittoneto committed:
Remove error of previous step

doudou commented:
@joaobrittoneto : is it ready to merge with these last changes ?

joaobrittoneto commented:
Yeap.
The last commit improved the simulation.
It still diverge of uwv_dynamic_model, but I start to think it is caused when the vehicle goes to a pitch near 90 deg. A pitch control may prevent it.
Anyways, the simulation is stable. Configuring it via sdf is straightforward also.

doudou commented:
Great work ! :+1:

doudou commented:
So, we should first add pitch control and only then add the added mass to the SDF model, right ?

joaobrittoneto commented:
Yes, I'd say so.
Other option is to force angular stability by setting a high center of buoyancy.
